### PR TITLE
Flow-ify ReactPartialRenderer

### LINKED ISSUE #11175: Flow-ify ReactPartialRenderer.js
It doesn’t have a `@flow` annotation. We should start Flow-ifying it to avoid bugs like https://github.com/facebook/react/pull/11173.

@iamdustan Wanna take it?

### PR DESCRIPTION
**Pull request by iamdustan**
Closes #11175 

cc @gaearon

### CODE REVIEW

File: src/renderers/shared/server/ReactPartialRenderer.js [hunk 1]
```diff
@@ -443,18 +445,33 @@ function resolve(child, context) {
   return {child, context};
 }
 
+type Frame = {
+  domNamespace: string,
+  children: any,
+  childIndex: number,
+  context: Object,
+  footer: string,
+  debugElementStack: Array<any>,
```
Comment (iamdustan) [inline src/renderers/shared/server/ReactPartialRenderer.js]: all the `Frame` annotations below only add `debugElementStack` in dev mode which causd a lot of errors in the logging utility functions. Just declaring it as always present (which it is in `dev`, and I think appropriately defended against in `prod`) solved a lot of the flow errors.

File: src/renderers/shared/server/ReactPartialRenderer.js [hunk 2]
```diff
@@ -320,6 +321,7 @@ function validateRenderResult(child, type) {
 
 function resolve(child, context) {
   while (React.isValidElement(child)) {
+    child = (child: React.Element);
```
Comment (gaearon) [inline src/renderers/shared/server/ReactPartialRenderer.js]: Why is this needed?

File: src/renderers/shared/server/ReactPartialRenderer.js [hunk 3]
```diff
@@ -443,18 +445,36 @@ function resolve(child, context) {
...
+    var topFrame: Frame = ({
       // Assume all trees start in the HTML namespace (not totally true, but
       // this is what we did historically)
       domNamespace: Namespaces.html,
       children,
       childIndex: 0,
       context: emptyObject,
       footer: '',
-    };
+    }: any);
```
Comment (gaearon) [inline src/renderers/shared/server/ReactPartialRenderer.js]: Why?

File: src/renderers/shared/server/ReactPartialRenderer.js [hunk 4]
```diff
@@ -525,13 +549,13 @@ class ReactDOMServerRenderer {
...
           var children = toArray(child);
-          var frame = {
+          var frame: Frame = ({
             domNamespace: parentNamespace,
             children,
             childIndex: 0,
             context: context,
             footer: '',
-          };
+          }: Object);
```
Comment (gaearon) [inline src/renderers/shared/server/ReactPartialRenderer.js]: Why?

File: src/renderers/shared/server/ReactPartialRenderer.js [hunk 5]
```diff
@@ -821,14 +845,14 @@ class ReactDOMServerRenderer {
...
-    var frame = {
+    var frame: Frame = ({
       domNamespace: getChildNamespace(parentNamespace, element.type),
       tag,
       children,
       childIndex: 0,
       context: context,
       footer: footer,
-    };
+    }: any);
```
Comment (gaearon) [inline src/renderers/shared/server/ReactPartialRenderer.js]: Why?

File: src/renderers/shared/server/ReactPartialRenderer.js (hunk 3 above)
Comment (iamdustan) [reply src/renderers/shared/server/ReactPartialRenderer.js]: all of these are because of https://github.com/facebook/react/pull/11251#discussion_r145140768

```jsx
var frame = {
  ...everythingExcept_debugElementStack
};
if (process.env.NODE_ENV) {
  frame.debugElementStack = [];
}
```

So the `frame: Frame = ({}: any)` annotation is so that flow knows that this object can have the `debugElementStack` property.

```jsx
var frame: Frame = ({
  ...everythingExcept_debugElementStack
// the any is to prevent Flow from considered this a
// sealed object without the debugElementStack
}: any);
```

without this then the following chunks of code have a lot of errors due to `debugElementStack` being `undefined`

https://github.com/iamdustan/react/blob/bacc1e72f58e1ec3eab16c52d6db28c7d0c21014/src/renderers/shared/server/ReactPartialRenderer.js#L65-L98

File: src/renderers/shared/server/ReactPartialRenderer.js (hunk 2 above)
Comment (iamdustan) [reply src/renderers/shared/server/ReactPartialRenderer.js]: I’m not really sure why Flow isn’t understanding that `child` is a `React.Element` at this point without this extra hint. I thought that by typing the callsites to `resolve` this would just work.

File: src/renderers/shared/server/ReactPartialRenderer.js (hunk 3 above)
Comment (gaearon) [reply src/renderers/shared/server/ReactPartialRenderer.js]: Typing something as `any` is generally as bad as not typing in the first place. Or possibly worse because now you have an illusion of type safety.

Could we perhaps use approach like

https://github.com/facebook/react/blob/90370f28ff1ce876c8ee6ec0062c921cf05b7924/src/renderers/dom/fiber/ReactDOMFiberEntry.js#L200-L201

and 

https://github.com/facebook/react/blob/90370f28ff1ce876c8ee6ec0062c921cf05b7924/src/renderers/dom/fiber/ReactDOMFiberEntry.js#L210

?

Or just treat that field as always optional?

Comment (iamdustan) [reply src/renderers/shared/server/ReactPartialRenderer.js]: It is typed similarly, just `var binding: Type = ({}: any);` instead of `var binding = (({}: any): Type);`

Though I’ll look to into make the making separate `*Prod` and `*Dev` variants yet.

File: packages/react-dom/src/server/ReactPartialRenderer.js [hunk 6]
```diff
@@ -319,6 +321,7 @@ function validateRenderResult(child, type) {
 
 function resolve(child, context) {
   while (React.isValidElement(child)) {
+    child = (child: React.Element);
```
Comment (gaearon) [inline packages/react-dom/src/server/ReactPartialRenderer.js]: This cast seemed suspicious to me. Because if it "worked" then Flow would've inferred it anyway.

I checked, and it seems like `React.Element` is `any` 😞 Flow can be so confusing. This is not the thing we're supposed to be using, apparently.

File: packages/react-dom/src/server/ReactPartialRenderer.js [hunk 7]
```diff
@@ -442,18 +445,36 @@ function resolve(child, context) {
...
+  context: Object,
+  footer: string,
+  debugElementStack: Array<React.Element>,
+};
+
 class ReactDOMServerRenderer {
-  constructor(element, makeStaticMarkup) {
+  stack: Array<Frame>;
+  exhausted: boolean;
+  currentSelectValue: null;
```
Comment (gaearon) [inline packages/react-dom/src/server/ReactPartialRenderer.js]: This is not a very helpful annotation. It's not always `null`. So by encoding it like this we actually make it harder to fix the Flow coverage because now it *looks* like fixing this is a Flow violation.

File: packages/react-dom/src/server/ReactPartialRenderer.js [hunk 8]
```diff
@@ -317,12 +328,20 @@ function validateRenderResult(child, type) {
...
+function resolve(
+  child: mixed,
+  context: Object,
+): {|
+  child: mixed,
+  context: Object,
+|} {
   while (React.isValidElement(child)) {
+    // Safe because we just checked it's an element.
+    var element: ReactElement = ((child: any): ReactElement);
```
[the remark this answers is not in the document]
Comment (iamdustan) [reply packages/react-dom/src/server/ReactPartialRenderer.js]: Do you need the `: ReactElement` on both sides of the assignment? I’ve only ever done one or the other and it always seemed to work as expected.

File: packages/react-dom/src/server/ReactPartialRenderer.js (hunk 6 above)
Comment (iamdustan) [reply packages/react-dom/src/server/ReactPartialRenderer.js]: Yeah, one of my few wishes for flow is that it would be easier to see what it thought something was at any point in time and why.

File: packages/react-dom/src/server/ReactPartialRenderer.js [hunk 9]
```diff
@@ -66,7 +66,8 @@ if (__DEV__) {
   var currentDebugStack = null;
   var currentDebugElementStack = null;
   var setCurrentDebugStack = function(stack) {
-    currentDebugElementStack = stack[stack.length - 1].debugElementStack;
+    var frame: Frame = stack[stack.length - 1];
```
Comment (iamdustan) [inline packages/react-dom/src/server/ReactPartialRenderer.js]: since this entire chunk of code is inside a `if (__DEV__)` block I think we could safely say that `var frame: FrameDev = stack[stack.length - 1];`

File: packages/react-dom/src/server/ReactPartialRenderer.js [hunk 10]
```diff
@@ -442,20 +461,46 @@ function resolve(child, context) {
...
       domNamespace: Namespaces.html,
-      children,
+      children: childrenArray,
       childIndex: 0,
       context: emptyObject,
       footer: '',
     };
     if (__DEV__) {
-      topFrame.debugElementStack = [];
+      ((topFrame: any): FrameDev).debugElementStack = [];
```
[the remark this answers is not in the document]
Comment (iamdustan) [reply packages/react-dom/src/server/ReactPartialRenderer.js]: I was trying (and unable) to find a way to get Flow to know that the only variant between `Frame` and `FrameDev` that could ever make it to the `*CurrentDebugStack` methods above were `FrameDev` types.

File: packages/react-dom/src/server/ReactPartialRenderer.js (hunk 8 above)
[the remark this answers is not in the document]
Comment (gaearon) [reply packages/react-dom/src/server/ReactPartialRenderer.js]: Not necessary. Can remove.

File: packages/react-dom/src/server/ReactPartialRenderer.js (hunk 10 above)
[the remark this answers is not in the document]
Comment (gaearon) [reply packages/react-dom/src/server/ReactPartialRenderer.js]: Not sure it's possible. I think using PROD type most things and then unsafely casting to DEV type for DEV things is the way to go.

File: packages/react-dom/src/server/ReactPartialRenderer.js (hunk 9 above)
Comment (gaearon) [reply packages/react-dom/src/server/ReactPartialRenderer.js]: I guess yea.
